microservices/collaboration-service/src/index.ts

```diff
@@ -0,0 +1,68 @@
+import { Server } from "socket.io";
+
+const io = new Server(3111, {
+  cors: {
+    origin: ["http://localhost:5173", "http://127.0.0.1:5173"],
+  },
+});
+
+io.use((socket, next) => {
+  console.log("Incoming connection");
+  const roomId = socket.handshake["query"]["roomId"];
+  if (roomId === undefined || Array.isArray(roomId)) {
+    next(new Error("Missing room id"));
+  }
+
+  next();
+});
+
+io.on("connection", (socket) => {
+  console.log(`Connection established with socket id ${socket.id}`);
+
+  // Client sends { query: { roomId: actualRoomId } }
+  // Middleware prevents undefined or array value so we typecast
+  const room = socket.handshake["query"]["roomId"] as string;
+
+  // Total rooms should be 1 for Peerprep
+  socket.join(room);
+  console.log(`A user joined room ${room}`);
+  console.log(`Total rooms: ${socket.rooms.size - 1}`);
+
+  // emit number of clients in room to ensure room is ready
+  io.in(room).emit("room count", io.sockets.adapter.rooms.get(room)?.size);
+
+  socket.on("client code changes", (change) => {
+    socket.to(room).emit("server code changes", change);
+  });
+
+  socket.on("other user has left", () => {
+    console.log("Connection terminated because other user has left");
+    socket.disconnect();
+  });
```

Comment:
Are we still disconnecting if other user leaves? Or are we going with our discussion on tuesday to remain connected as it there will be less state to maintain

Reply:
I will merge first, randall has changes which will be applied on top when he submit his PR